plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergReadVersionedTable.java

```diff
@@ -0,0 +1,138 @@
+        assertQuery("SELECT * FROM test_schema.test_iceberg_read_versioned_table FOR TIMESTAMP AS OF " + timestamp(v1Unixtime, 8), "VALUES ('a', 1)");
+        assertQuery("SELECT * FROM test_schema.test_iceberg_read_versioned_table FOR TIMESTAMP AS OF " + timestamp(v2Unixtime, 8), "VALUES ('a', 1), ('b', 2)");
+    }
+
+    @Test
+    public void testSelectTableWithEndShortTimestampWithTimezone()
+    {
+        assertQueryFails("SELECT * FROM test_schema.test_iceberg_read_versioned_table FOR TIMESTAMP AS OF " + shortTzTimestamp(1),
+                format("Time travel data is not available for table %s. The requested time is beyond the allowed time travel period.", "\"test_schema\".\"test_iceberg_read_versioned_table\""));
+        assertQuery("SELECT * FROM test_schema.test_iceberg_read_versioned_table FOR TIMESTAMP AS OF " + shortTzTimestamp(v1Unixtime), "VALUES ('a', 1)");
+        assertQuery("SELECT * FROM test_schema.test_iceberg_read_versioned_table FOR TIMESTAMP AS OF " + shortTzTimestamp(v2Unixtime), "VALUES ('a', 1), ('b', 2)");
+    }
+
+    @Test
+    public void testSelectTableWithEndLongTimestampWithTimezone()
+    {
+        assertQueryFails("SELECT * FROM test_schema.test_iceberg_read_versioned_table FOR TIMESTAMP AS OF " + longTzTimestamp(1),
+                format("Time travel data is not available for table %s. The requested time is beyond the allowed time travel period.", "\"test_schema\".\"test_iceberg_read_versioned_table\""));
+        assertQuery("SELECT * FROM test_schema.test_iceberg_read_versioned_table FOR TIMESTAMP AS OF " + longTzTimestamp(v1Unixtime), "VALUES ('a', 1)");
+        assertQuery("SELECT * FROM test_schema.test_iceberg_read_versioned_table FOR TIMESTAMP AS OF " + longTzTimestamp(v2Unixtime), "VALUES ('a', 1), ('b', 2)");
+    }
+
+    @Test
+    public void testEndVersionInTableNameAndForClauseShouldFail()
+    {
+        assertQueryFails("SELECT * FROM test_schema.\"test_iceberg_read_versioned_table@" + v1SnapshotId + "\" FOR VERSION AS OF " + v1SnapshotId,
+                "Cannot specify end version both in table name and FOR clause");
+
+        assertQueryFails("SELECT * FROM test_schema.\"test_iceberg_read_versioned_table@" + v1SnapshotId + "\" FOR TIMESTAMP AS OF " + longTzTimestamp(v1Unixtime),
+                "Cannot specify end version both in table name and FOR clause");
+    }
+
+    private String shortTzTimestamp(long unixtime)
+    {
+        return "from_unixtime(" + unixtime + ")";
+    }
+
+    private String longTzTimestamp(long unixtime)
+    {
+        return "from_unixtime_nanos(" + unixtime + "000000000)";
+    }
+
+    private String timestamp(long unixtime, int precision)
+    {
+        return "CAST(" + shortTzTimestamp(unixtime) + " AT TIME ZONE 'UTC' AS TIMESTAMP(" + precision + ") WITHOUT TIME ZONE)";
+    }
+
+    private Optional<Long> getLatestSnapshotId(String schemaName, String tableName)
```

Comment:
Use the pattern from `io.trino.plugin.iceberg.BaseIcebergConnectorTest#getLatestSnapshotId` or extract an utility method